test/test_kernel_launch.jl

```diff
@@ -0,0 +1,138 @@
+    if !compatible(backend, T)
+        continue
+    end
+
+    @testset "$(basename(@__FILE__)) (backend: $backend, type: $T)" begin
+        if (backend isa CPU)
+            @testset "deepmap! on CPU backend" begin
+                # struct CuArray end # in reality this is replaced by "using CUDA"
+
+                calls = Symbol[]
+
+                disable_task_sync!(::Any) = nothing
+                enable_task_sync!(::Any) = nothing
+                disable_task_sync!(::CuArray) = push!(calls, :disable) # actual unsafe_disable_task_sync! call
+                enable_task_sync!(::CuArray) = push!(calls, :enable) # actual unsafe_enable_task_sync! call
+
+                empty!(calls)
+                foo1 = (1, 2, 3)
+                deepmap!(disable_task_sync!, foo1) # prints nothing
+                deepmap!(enable_task_sync!, foo1)  # prints nothing
+                @test isempty(calls)
+
+                empty!(calls)
+                foo2 = ()
+                deepmap!(disable_task_sync!, foo2) # prints nothing
+                deepmap!(enable_task_sync!, foo2)  # prints nothing
+                @test isempty(calls)
+
+                empty!(calls)
+                foo3 = (CuArray(), CuArray(), (1, ()))
+                deepmap!(disable_task_sync!, foo3) # prints "disable" twice
+                deepmap!(enable_task_sync!, foo3) # prints "enable" twice
+                @test calls == [:disable, :disable, :enable, :enable]
+
+                empty!(calls)
+                foo4 = (a=1, b=CuArray(), c=((), 1))
+                deepmap!(disable_task_sync!, foo4) # prints "disable" once
+                deepmap!(enable_task_sync!, foo4) # prints "enable" once
+                @test calls == [:disable, :enable]
+
+                empty!(calls)
+                struct Foo{T}
+                    a::Int
+                    b::CuArray
+                    c::T
+                end
+                foo5 = Foo(1, CuArray(), foo4)
+
+                with_no_task_sync!(foo5) do
+                    push!(calls, :in_out) # kernel execution for inner and outer regions goes here
+                end # will print "disable" twice, then "in_out" and "enable" twice
+                @show calls
+                # @test calls == [:disable, :disable, :in_out, :enable, :enable]
+            end
+        end
+
+        @testset "deepmap! on Field" begin
+            count = DataType[]
+
+            arch = Arch(backend)
+            grid = UniformGrid(arch; origin=(T(0.0), T(0.0)), extent=(T(1.0), T(1.0)), dims=(6, 4))
+
+            Δt = T(1.0)
+            C  = Field(backend, grid, Center())
+            q  = VectorField(backend, grid)
+
+            M = typeof(C.data)
+            myfun!(x::M) where {M} = push!(count, typeof(x))
+            # myfun!(x::M) = push!(count, typeof(x))
+
+            @testset "deepmap! on single Field" begin
+                empty!(count)
+                deepmap!(myfun!, C)
+                @test length(count) == 1
```

Comment:
Looks like `myfun!` does not dispatch on the correct type and thus count gets filled with the `myfun(::Any)` method